### PR TITLE
STY/ENH: Testing Instruments -- metadata and start_time

### PR DESCRIPTION
**Pull request by jklenzing**
# Description

Addresses #389 (partial) and #621

Moves the metadata for the test instruments to a single location.  All possible values are generated, but only the values required are passed through to the instrument.  Expanded some values of metadata to include min and max values.

Adds `start_time` to the test instruments as an optional kwarg.  If specified, the start time will be adjusted by the specified offset (as a `dt.timedelta`).
- Includes unit tests for the changes to `generate_times` as part of new test_methods_testing.py file.
- Includes integration tests for each test instrument to ensure both `start_time` and `num_samples` are passed through correctly.
  - Note that end time is not tested in the integration tests since each instrument has a different hard-wired frequency.  However, this is captured in the unit tests on the function itself.
 - Existing behavior that times will not be generated on the next day is maintained (ie, num_samples=87000 points will still only generate 86400 points at a 1-sec cadence).

Caught a bug: index was mistakenly added as 'uts' in the two orbital xarray objects.

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

By loading each test instrument and inspecting metadata, and via pytest.

**Test Configuration**:
* Operating system: Mac  10.15.7
* Version number: Python 3.8.2

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

If this is a release PR, replace the first item of the above checklist with the release
checklist on the wiki: https://github.com/pysat/pysat/wiki/Checklist-for-Release

### CODE REVIEW

File: pysat/instruments/methods/testing.py [hunk 1]
```diff
@@ -87,6 +89,104 @@ def preprocess(self, test_preprocess_kwarg=None):
...
+
+    Parameters
+    ----------
+    epoch_name : str
+        The variable name of the instrument epoch.
+    data : pds.DataFrame or xr.Dataset
+        The dataset keys from the instrument.
+
+    """
+    # create standard metadata for all parameters
```
Comment (aburrell) [inline pysat/instruments/methods/testing.py:106]:
```suggestion
    # Create standard metadata for all parameters
```

File: pysat/instruments/pysat_testing.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
         'no_download': 'simulate an instrument without download support',
         'non_strict': 'simulate an instrument without strict_time_flag',
         'user_password': 'simulates an instrument that requires a password',
-        'default_meta': 'simulates an instrument using the defualt meta'}
+        'default_meta': 'simulates an instrument using the default meta'}
```
Comment (aburrell) [inline pysat/instruments/pysat_testing.py:25]:
```suggestion
        'default_meta': 'simulates an instrument using the default metadata'}
```

File: pysat/instruments/pysat_testing.py [hunk 3]
```diff
@@ -45,7 +45,7 @@
 
 def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
          sim_multi_file_left=False, root_date=None, malformed_index=False,
-         num_samples=None, test_load_kwarg=None):
+         start_time=None, num_samples=None, test_load_kwarg=None):
```
Comment (aburrell) [inline pysat/instruments/pysat_testing.py:48]:
```suggestion
         start_time=None, num_samples=86400, test_load_kwarg=None):
```

File: pysat/instruments/pysat_testing.py [hunk 4]
```diff
@@ -68,8 +68,13 @@ def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
...
     malformed_index : boolean
         If True, time index will be non-unique and non-monotonic (default=False)
+    start_time : dt.timedelta or NoneType
+        Offset time of start time since midnight UT. If None, instrument data
+        will begin at midnight.
+        (default=None)
     num_samples : int
-        Number of samples per day
+        Maximum number of times to generate.  Data points will not go beyond the
+        current day.
```
Comment (aburrell) [inline pysat/instruments/pysat_testing.py:77]:
```suggestion
        current day. (default=86400)
```

File: pysat/instruments/pysat_testing.py [hunk 5]
```diff
@@ -92,8 +97,8 @@ def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
     if num_samples is None:
         # Default to 1 day at a frequency of 1S
         num_samples = 86400
```
Comment (aburrell) [inline pysat/instruments/pysat_testing.py:99]:
```suggestion
```

File: pysat/instruments/pysat_testing2d.py [hunk 6]
```diff
@@ -32,7 +32,7 @@
 
 
 def load(fnames, tag=None, inst_id=None, malformed_index=False,
-         num_samples=None, test_load_kwarg=None):
+         start_time=None, num_samples=None, test_load_kwarg=None):
```
Comment (aburrell) [inline pysat/instruments/pysat_testing2d.py:35]:
```suggestion
         start_time=None, num_samples=864, test_load_kwarg=None):
```

File: pysat/instruments/pysat_testing2d.py [hunk 7]
```diff
@@ -46,8 +46,13 @@ def load(fnames, tag=None, inst_id=None, malformed_index=False,
...
         If True, the time index will be non-unique and non-monotonic.
         (default=False)
+    start_time : dt.timedelta or NoneType
+        Offset time of start time since midnight UT. If None, instrument data
+        will begin at midnight.
+        (default=None)
     num_samples : int
-        Number of samples
+        Maximum number of times to generate.  Data points will not go beyond the
+        current day.
```
Comment (aburrell) [inline pysat/instruments/pysat_testing2d.py:55]:
```suggestion
        current day. (default=864)
```

File: pysat/instruments/pysat_testing2d.py [hunk 8]
```diff
@@ -71,8 +76,8 @@ def load(fnames, tag=None, inst_id=None, malformed_index=False,
         num_samples = 864
```
Comment (aburrell) [inline pysat/instruments/pysat_testing2d.py:76]: Remove this `if` statement.

File: pysat/instruments/pysat_testing2d.py [hunk 9]
```diff
@@ -71,8 +76,8 @@ def load(fnames, tag=None, inst_id=None, malformed_index=False,
...
 
     # Using 100s frequency for compatibility with seasonal analysis unit tests
-    uts, index, dates = mm_test.generate_times(fnames, num_samples,
-                                               freq='100S')
+    uts, index, dates = mm_test.generate_times(fnames, num_samples, freq='100S',
+                                               start_time=start_time)
     # seed DataFrame with UT array
```
Comment (aburrell) [inline pysat/instruments/pysat_testing2d.py:81]:
```suggestion
    # Seed the DataFrame with a UT array
```

File: pysat/instruments/pysat_testing2d_xarray.py [hunk 10]
```diff
@@ -38,7 +38,7 @@
 
 
 def load(fnames, tag=None, inst_id=None, malformed_index=False,
-         num_samples=None, test_load_kwarg=None):
+         start_time=None, num_samples=None, test_load_kwarg=None):
```
Comment (aburrell) [inline pysat/instruments/pysat_testing2d_xarray.py:41]:
```suggestion
         start_time=None, num_samples=864, test_load_kwarg=None):
```

File: pysat/instruments/pysat_testing2d_xarray.py [hunk 11]
```diff
@@ -51,8 +51,13 @@ def load(fnames, tag=None, inst_id=None, malformed_index=False,
...
     malformed_index : bool False
         If True, the time index will be non-unique and non-monotonic.
+    start_time : dt.timedelta or NoneType
+        Offset time of start time since midnight UT. If None, instrument data
+        will begin at midnight.
+        (default=None)
     num_samples : int
-        Number of samples
+        Maximum number of times to generate.  Data points will not go beyond the
+        current day.
```
Comment (aburrell) [inline pysat/instruments/pysat_testing2d_xarray.py:60]:
```suggestion
        current day. (default=864)
```

File: pysat/instruments/pysat_testing2d_xarray.py [hunk 12]
```diff
@@ -76,16 +81,16 @@ def load(fnames, tag=None, inst_id=None, malformed_index=False,
         # Default to 1 day at a frequency of 100S
         num_samples = 864
```
Comment (aburrell) [inline pysat/instruments/pysat_testing2d_xarray.py:82]: Remove

File: pysat/instruments/pysat_testing_xarray.py [hunk 13]
```diff
@@ -39,7 +39,7 @@
 
 def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
          sim_multi_file_left=False, malformed_index=False,
-         num_samples=None, test_load_kwarg=None):
+         start_time=None, num_samples=None, test_load_kwarg=None):
```
Comment (aburrell) [inline pysat/instruments/pysat_testing_xarray.py:42]:
```suggestion
         start_time=None, num_samples=86400, test_load_kwarg=None):
```

File: pysat/instruments/pysat_testing_xarray.py [hunk 14]
```diff
@@ -58,8 +58,13 @@ def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
...
     malformed_index : boolean
         If True, time index will be non-unique and non-monotonic.
+    start_time : dt.timedelta or NoneType
+        Offset time of start time since midnight UT. If None, instrument data
+        will begin at midnight.
+        (default=None)
     num_samples : int
-        Number of samples
+        Maximum number of times to generate.  Data points will not go beyond the
+        current day.
```
Comment (aburrell) [inline pysat/instruments/pysat_testing_xarray.py:67]:
```suggestion
        current day. (default=86400)
```

File: pysat/instruments/pysat_testing_xarray.py [hunk 15]
```diff
@@ -82,8 +87,8 @@ def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
     if num_samples is None:
         # Default to 1 day at a frequency of 1S
         num_samples = 86400
```
Comment (aburrell) [inline pysat/instruments/pysat_testing_xarray.py:89]:
```suggestion
```

File: pysat/instruments/pysat_testmodel.py [hunk 16]
```diff
@@ -32,7 +32,7 @@
 preprocess = mm_test.preprocess
 
 
-def load(fnames, tag=None, inst_id=None, num_samples=None,
+def load(fnames, tag=None, inst_id=None, start_time=None, num_samples=None,
```
Comment (aburrell) [inline pysat/instruments/pysat_testmodel.py:35]:
```suggestion
def load(fnames, tag=None, inst_id=None, start_time=None, num_samples=96,
```

File: pysat/instruments/pysat_testmodel.py [hunk 17]
```diff
@@ -44,8 +44,13 @@ def load(fnames, tag=None, inst_id=None, num_samples=None,
...
     inst_id : str or NoneType
         Instrument satellite ID (accepts '')
+    start_time : dt.timedelta or NoneType
+        Offset time of start time since midnight UT. If None, instrument data
+        will begin at midnight.
+        (default=None)
     num_samples : int
-        Number of samples
+        Maximum number of times to generate.  Data points will not go beyond the
+        current day.
```
Comment (aburrell) [inline pysat/instruments/pysat_testmodel.py:53]:
```suggestion
        current day. (default=96)
```

File: pysat/instruments/pysat_testmodel.py [hunk 18]
```diff
@@ -65,8 +70,8 @@ def load(fnames, tag=None, inst_id=None, num_samples=None,
         # Default to 1 day at a frequency of 900S
         num_samples = 96
```
Comment (aburrell) [inline pysat/instruments/pysat_testmodel.py:71]: Remove

File: pysat/tests/test_instruments.py [hunk 19]
```diff
@@ -91,3 +94,38 @@ def teardown_class(self):
         return
```
Comment (aburrell) [inline pysat/tests/test_instruments.py:94]: On line 73 the docstring is missing the empty line before closing.

**Review by aburrell — CHANGES_REQUESTED** [review]

**Comment by jklenzing** [comment]
Thanks @aburrell! It's a lot cleaner now.

File: pysat/instruments/pysat_testmodel.py [hunk 20]
```diff
@@ -61,12 +66,9 @@ def load(fnames, tag=None, inst_id=None, num_samples=None,
     # Support keyword testing
     logger.info(''.join(('test_load_kwarg = ', str(test_load_kwarg))))
 
-    if num_samples is None:
-        # Default to 1 day at a frequency of 900S
-        num_samples = 96
-    # create an artifical satellite data set
-    uts, index, dates = mm_test.generate_times(fnames, num_samples,
-                                               freq='900S')
+    # create an artifical model data set
```
Comment (aburrell) [inline pysat/instruments/pysat_testmodel.py:69]:
```suggestion
    # Create an artificial model data set
```

**Review by aburrell — APPROVED** [review]
Found a small typo, would be easiest to fix now.

**Pull request merged**